### PR TITLE
Gulp

### PR DESCRIPTION
**Pull request by jimmynotjim**
With the move to a src/dist setup, Grunt tasks have become increasingly lengthy. Now that we're considering running all tests within Grunt, execution time will only increase because it runs each task in sequence (see note 3). In addition to these points, many of the plugins that would make this easier (like changed/newer) have either been abandoned or are wicked buggy. This PR is an exploration into using Gulp and how the two compare.

## Additions

- gulp and the required npm plugins
- gulp config file to lay out configs for each task
- gulp tasks split up into their own files

## Removals

- Removed Grunt plugins from package.json
- Left the Gruntfile to make sure we've covered everything

## Changes

- Updated setup.sh to use gulp ~~(see note 1)~~
- Updated travis to use gulp tasks
- Updated main.less to use the paths option in less compiler
- base.html to include listener for browser reloading (see note 2)

## Testing

- globally install Gulp: `npm install gulp -g`
- checkout branch, run `./setup.sh` ~~(see note 1)~~
- ~~update core imports of  `/src/vendor/capital-framework/capital-framework.less` (L:5999) from:~~
  __This has been fixed with the move the cf-v1, please ignore__
  ```less
  @import (less) "normalize.css";
  @import (less) "normalize-legacy-addon.css";
  @import (less) "cf-brand-colors.less";
  @import (less) "cf-vars.less";
  @import (less) "cf-media-queries.less";
  @import (less) "cf-utilities.less";
  @import (less) "cf-base.less";
  ```
  to:

  ```less
  @import (less) "normalize-css/normalize.css";
  @import (less) "normalize-legacy-addon/normalize-legacy-addon.css";
  @import (less) "cf-core/src/cf-brand-colors.less";
  @import (less) "cf-core/src/cf-vars.less";
  @import (less) "cf-core/src/cf-media-queries.less";
  @import (less) "cf-core/src/cf-utilities.less";
  @import (less) "cf-core/src/cf-base.less";
  ```
  ~~__This won't be needed once we fix our CF import process__~~
- ~~run `gulp build`~~
- if you want to use watching/browser reloading run `gulp watch` and refresh the browser

## Review

- @KimberlyMunoz 
- @anselmbradford 
- @sebworks 
- @contolini

[Preview this PR without the whitespace changes](?w=0)

## Notes

1. ~~Due to how we currently import CF, the `styles` task will fail without making the change listed under testing~~ _Fixed_
2. This listener is required to use the `watch` task, and I plan to explore ways to conditionally load it locally only
3. Times - Gulp is able to run tasks concurrently, so the biggest win is the build task, especially when you add in the other tests

  Task | Grunt | Gulp
  -----|-------|-----
  lint js | 3.9s | 1.81s
  styles (sourcemaps takes forever) |  2.6s | 2.55s
  scripts | 4.92s | 2.04s
  copy | 998ms | 524ms
  unit tests (only one in grunt) | 16.5s | 944ms
  build  | 21.8s | 11.3s
  
  I timed the gulp build task by hand because the concurrent tasks output how long they each took, but the entire build task doesn't tell you how long it took from start to finish.
4. This PR is as close as I could get to a 1:1 comparison with our existing build process
5. I included the browser, macro, and processor tests as part of the exploration in #780

### CODE REVIEW

**Comment by jimmynotjim** [comment]
Going to take this out of WIP for a hopeful full review by Wed (if we're going to use it, we should get it in early in the sprint).

File: gulpfile.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
+'use strict';
+
+/*
+  gulpfile.js
+  ===========
+  Rather than manage one giant configuration file responsible
```
Comment (anselmbradford) [inline gulpfile.js]: Minor: This first sentence (before comma) is probably not necessary.

File: gulpfile.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  for creating multiple tasks, each task has been broken out into
+  its own file in gulp/tasks. Any files in that directory get
+  automatically required below.
+  To add a new task, simply add a new task file that directory.
```
Comment (anselmbradford) [inline gulpfile.js]: Minor: `that directory` --> `in the gulp/tasks directory`.

File: gulp/tasks/clean.js [hunk 3]
```diff
@@ -0,0 +1,9 @@
+'use strict';
+
+var gulp = require( 'gulp' );
+var del = require( 'del' );
+var config = require( '../config' ).clean;
```
Comment (anselmbradford) [inline gulp/tasks/clean.js]: name `var configClean` or something to differentiate from `var config = require( '../config' );`

File: gulp/config.js [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    },
+    legacy: {
+      src:  loc.src + '/static-legacy/**/*',
+      dest: loc.dist + '/static-legacy'
+    },
+    icons: {
+      src:  loc.lib + '/cf-icons/src/fonts/*',
+      dest: loc.dist + '/static/fonts/'
+    },
+    pdfcss: {
```
Comment (anselmbradford) [inline gulp/config.js]: Consider merging `pdfcss` into the `vendorcss` task and renaming `pdffonts` to be `vendorfonts`.

File: gulp/tasks/copy.js [hunk 5]
```diff
@@ -0,0 +1,105 @@
+'use strict';
+
+var gulp = require( 'gulp' );
+var changed = require( 'gulp-changed' );
+var replace = require( 'gulp-replace' );
+var config = require( '../config' ).copy;
```
Comment (anselmbradford) [inline gulp/tasks/copy.js]: name `var configCopy` or something to differentiate from `var config = require( '../config' );`

File: gulp/tasks/copy.js [hunk 6]
```diff
@@ -0,0 +1,105 @@
+'use strict';
+
+var gulp = require( 'gulp' );
+var changed = require( 'gulp-changed' );
+var replace = require( 'gulp-replace' );
```
Comment (anselmbradford) [inline gulp/tasks/copy.js]: L:4 and L:5 technically should be `gulpChanged` and `gulpReplace`, but perhaps that gets too wordy in the implementation.

**Comment by anselmbradford** [comment]
Looks like https://github.com/cfpb/cfgov-refresh/commit/87de5155b44dc04999411cafdca90057beff52dd needs `gulp-replace` in `package.json` ... I'm getting `Error: Cannot find module 'gulp-replace'` on the latest update.

File: gulp/tasks/copy.js (hunk 6 above)
Comment (jimmynotjim) [inline gulp/tasks/copy.js]: I'm actually considering using the gulp auto-import plugin which would remove the requires anyway. Unless anyone has objections.

Comment (anselmbradford) [inline gulp/tasks/copy.js]: Sounds good. I don't know anything about it :smile: What're the trade-offs?

Comment (jimmynotjim) [inline gulp/tasks/copy.js]: Lazy loads the plugins as they're needed and reduces the `require()`s you need to keep up in your files https://www.npmjs.com/package/gulp-load-plugins

File: gulp/tasks/copy.js (hunk 5 above)
Comment (jimmynotjim) [inline gulp/tasks/copy.js]: Is that necessary? You're already in the `copy.js` file, it's kinda obvious what the `config` var is pointing to right?

**Comment by jimmynotjim** [comment]
> Looks like 87de515 needs gulp-replace

This was taken care of.

File: gulp/tasks/test.js [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    .on( 'finish', function() {
+      gulp.src( config.tests + '/unit_tests/**/*.js' )
+        .pipe( $.mocha( {
+          reporter: 'nyan'
+        } ) )
+        .pipe( $.istanbul.writeReports( {
+          dir: config.tests + '/unit_test_coverage'
+        } ) )
+
+        /* we want this but it breaks because we don't have good coverage
```
Comment (KimberlyMunoz) [inline gulp/tasks/test.js]: #realtalk

Comment (contolini) [inline gulp/tasks/test.js]: giphy-7

Comment (jimmynotjim) [inline gulp/tasks/test.js]: I was sad commenting it out, but I want to keep it in here so we don't forget about it

File: src/_layouts/base.html [hunk 8]
```diff
@@ -194,6 +194,11 @@
 <script src="{{ url_for('static', filename='js/routes/common.min.js') }}"></script>
 {% endblock javascript %}
 
+<script type='text/javascript' id="__bs_script__">
```
Comment (KimberlyMunoz) [inline src/_layouts/base.html]: Will this have any impact on the hosted version of flapjack?

Comment (jimmynotjim) [inline src/_layouts/base.html]: Ah, thanks for the reminder. I have a semi-working solution to keeping this local and I'll get it finished now.

File: gulp/tasks/test.js (hunk 7 above)
Comment (anselmbradford) [inline gulp/tasks/test.js]: `TODO: `

Comment (jimmynotjim) [inline gulp/tasks/test.js]: @anselmbradford updated

File: gulp/tasks/copy.js [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+  return gulp.src( config.vendorimg.src )
+    .pipe( $.changed( config.vendorimg.dest ) )
+    .on( 'error', handleErrors )
+    .pipe( gulp.dest( config.vendorimg.dest ) )
+    .pipe( browserSync.reload( {
+      stream: true
+    } ) );
+} );
+
+gulp.task( 'copy:vendorjs', function() {
```
Comment (sebworks) [inline gulp/tasks/copy.js]: Is it possible to consolidate the functions so that they call one task and pass in the config variable?

Comment (jimmynotjim) [inline gulp/tasks/copy.js]: Unfortunately no, because each has a different source and destination, which Gulp isn't really up to handling unless you're globbing them together (tried it and broke everything). It did kind of work for some files in Grunt, but that `process` task was super complicated and fragile. Breaking them each out here seems simpler to maintain and update when deps change.

**Comment by KimberlyMunoz** [comment]
LGTM

File: gulp/tasks/copy.js (hunk 9 above)
Comment (sebworks) [inline gulp/tasks/copy.js]: @jimmynotjim  I was thinking ( I think your approach is better for the reasons you mentioned ) :

```
function copyFiles ( src, destination ) {

  return gulp.src( src )
    .pipe( changed( destination ) )
    .on( 'error', handleErrors )
    .pipe( gulp.dest( destination ) )
    .pipe( browserSync.reload( {
      stream: true
    } ) );

}

gulp.task( 'copy:files', copyFiles.bind( undefined, config.files.src, config.files.dest ) );

gulp.task( 'copy:legacy', copyFiles.bind( undefined, config.legacy.src, config.legacy.dest ) );

```

**Comment by anselmbradford** [comment]
Works on 7001! :+1:

**Comment by jimmynotjim** [comment]
Rebased and waiting on travis to pass to be sure we're good

https://www.youtube.com/watch?v=-W6as8oVcuM

**Comment by jimmynotjim** [comment]
Conflicts! Fixed, no one touch anything!
